lib/libzfs/libzfs_crypto.c

```diff
@@ -0,0 +1,1594 @@
+	switch (keyloc) {
+	case ZFS_KEYLOCATION_PROMPT:
+		fd = stdin;
+		if (isatty(fileno(fd))) {
+			can_retry = B_TRUE;
+
+			/* raw keys cannot be entered on the terminal */
+			if (keyformat == ZFS_KEYFORMAT_RAW) {
+				ret = EINVAL;
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "Cannot enter raw keys on the terminal"));
+				goto error;
+			}
+		}
+		break;
+	case ZFS_KEYLOCATION_URI:
+		fd = fopen(&keylocation[7], "r");
+		if (!fd) {
+			ret = errno;
+			errno = 0;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Failed to open key material file"));
+			goto error;
+		}
+		break;
+	default:
+		ret = EINVAL;
+		zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+		    "Invalid keylocation."));
+		goto error;
+	}
+
+	/* fetch the key material into the buffer */
+	ret = get_key_material_raw(fd, fsname, keyformat, B_FALSE, newkey,
+	    &km, &kmlen);
+	if (ret != 0)
+		goto error;
+
+	/* do basic validation of the key material */
+	switch (keyformat) {
+	case ZFS_KEYFORMAT_RAW:
+		/* verify the key length is correct */
+		if (kmlen < WRAPPING_KEY_LEN) {
+			ret = EINVAL;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Raw key too short (expected %u)."),
+			    WRAPPING_KEY_LEN);
+			goto error;
+		}
+
+		if (kmlen > WRAPPING_KEY_LEN) {
+			ret = EINVAL;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Raw key too long (expected %u)."),
+			    WRAPPING_KEY_LEN);
+			goto error;
+		}
+		break;
+	case ZFS_KEYFORMAT_HEX:
+		/* verify the key length is correct */
+		if (kmlen < WRAPPING_KEY_LEN * 2) {
+			ret = EINVAL;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Hex key too short (expected %u)."),
+			    WRAPPING_KEY_LEN * 2);
+			goto error;
+		}
+
+		if (kmlen > WRAPPING_KEY_LEN * 2) {
+			ret = EINVAL;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Hex key too long (expected %u)."),
+			    WRAPPING_KEY_LEN * 2);
+			goto error;
+		}
+
+		/* check for invalid hex digits */
+		for (i = 0; i < WRAPPING_KEY_LEN * 2; i++) {
+			if (!isxdigit((char)km[i])) {
+				ret = EINVAL;
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "Invalid hex character detected."));
+				goto error;
+			}
+		}
+		break;
+	case ZFS_KEYFORMAT_PASSPHRASE:
+		/* verify the length is within bounds */
+		if (kmlen > MAX_PASSPHRASE_LEN) {
+			ret = EINVAL;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Passphrase too long (max %u)."),
+			    MAX_PASSPHRASE_LEN);
+			goto error;
+		}
+
+		if (kmlen < MIN_PASSPHRASE_LEN) {
+			ret = EINVAL;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Passphrase too short (min %u)."),
+			    MIN_PASSPHRASE_LEN);
+			goto error;
+		}
+		break;
+	default:
+		/* can't happen, checked above */
+		break;
+	}
+
+	if (do_verify && isatty(fileno(fd))) {
+		ret = get_key_material_raw(fd, fsname, keyformat, B_TRUE,
+		    newkey, &km2, &kmlen2);
+		if (ret != 0)
+			goto error;
+
+		if (kmlen2 != kmlen ||
+		    (memcmp((char *)km, (char *)km2, kmlen) != 0)) {
```

Comment:
I don't think that's necessary here. This just checks that you entered the same key twice in a row when creating a dataset or changing your existing key. It's not really an attack vector because the user needs to have the key already (or be creating it themselves) at this time.